### DISCUSSION TITLE
MapFileWriter wayDataBuffer too small for my application

### DISCUSSION THREAD

**Discussion by utack** [mapsforge/mapsforge#1509] [Q&A] [answered]
Hello,

I am coming from the wahooMapsCreator and I am running into an error that seems to be rooted in your MapFileWriter (via osmosis).
Long story short: I am running into an issue that [WAY_DATA_BUFFER_SIZE](https://github.com/mapsforge/mapsforge/blob/2650664ca7d4dde66c6791434fbaf9173256f4c2/mapsforge-map-writer/src/main/java/org/mapsforge/map/writer/MapFileWriter.java#L278) seems to be too small for my use case

`SEVERE: Thread for task 1-rb failed
java.nio.BufferOverflowException
	at java.base/java.nio.HeapByteBuffer.put(HeapByteBuffer.java:238)
	at org.mapsforge.map.writer.MapFileWriter.processTile(MapFileWriter.java:902)
	at org.mapsforge.map.writer.MapFileWriter.writeSubfile(MapFileWriter.java:977)
	at org.mapsforge.map.writer.MapFileWriter.writeFile(MapFileWriter.java:375)
	at org.mapsforge.map.writer.osmosis.MapFileWriterTask.complete(MapFileWriterTask.java:113)
	at crosby.binary.osmosis.OsmosisBinaryParser.complete(OsmosisBinaryParser.java:33)
	at crosby.binary.file.BlockInputStream.process(BlockInputStream.java:38)
	at crosby.binary.osmosis.OsmosisReader.run(OsmosisReader.java:46)
	at java.base/java.lang.Thread.run(Thread.java:1570)
`

I am not sure if this is a bug, since the size is fixed, is it expected to occur in some cases?
I have a lot of tags in the file and as such a lot of data.
Let me know if you need the files and command line to reproduce this, but it seems pretty straightforward.

Thank you for your help

**Comment by devemux86** [comment]
@utack thanks for the report.

There are no problems all the time we create maps in [Mapsforge server](https://download.mapsforge.org/)
or from the other map providers like [OpenAndroMaps](https://www.openandromaps.org/en), [Freizeitkarte](https://www.freizeitkarte-osm.de/android/en/index.html), etc.

You can change the size in the code, rebuild the jar and try it in your maps?
Also the maximum number of tags per element is **15** (https://github.com/mapsforge/mapsforge/issues/971).

**Comment by utack** [comment]
oh that might be it...I am using waaay more tags than that

as far as changing the size goes: unfortunately I do not have the experience, I managed to build a jar but osmosis complained the "--mw" command is not defined with it

Maybe you can help me @devemux86 ?

I added

`plugins {
    id 'com.github.johnrengelman.shadow' version '8.1.1'
}
`

to the build.gradle and ran 

`ANDROID_HOME=/tmp grade shadowJar
`

which produced the non-functioning jar

**Comment by utack** [comment]
I managed to build it, my oversight that there is a "farJar" task in the grade.build

The result is the same using 0xFFFFFFF as 

Can I hand you the source files and commandline to reproduce this issue?

**Comment by devemux86** [comment]
Need to follow the specification and use up to 15 tags per element.
Then see if the map creation works.

**Comment by utack** [comment]
> Need to follow the specification and use up to 15 tags per element. Then see if the map creation works.

Yes, that was the first thing I fixed  
Did not change things unfortunately!

**Comment by devemux86** [comment]
Can download a small .pbf from [Geofabrik](https://download.geofabrik.de/) and try creating a simple map.
There are detailed instructions in the [map creation guide](https://github.com/mapsforge/mapsforge/blob/master/docs/MapCreation.md).

Once this is working, then can start creating custom maps.
Custom maps usually require a new custom [tag-mapping](https://github.com/mapsforge/mapsforge/blob/master/docs/Getting-Started-Map-Writer.md#defining-a-custom-tag-mapping-via-xml).

**Comment by utack** [comment]
All the other "tiles" in the wahoomc proccess work and prioduce a "map" file, it is only 1/80 tiles that fails
So the setup seems solid
I can also change my custom tag-mapping to produce or avoid the error, by removing some ways of type "building"

**Comment by applantation** [comment]
You could also preprocess the data to reduce the number of building types -
particularly if the specific type does not matter for rendering

Jakob ***@***.***> schrieb am So. 28. Juli 2024 um 13:41:

**Comment by devemux86** [comment]
[TagTransform](https://wiki.openstreetmap.org/wiki/Osmosis/TagTransform) is an Osmosis plugin useful in such cases.

**Comment by devemux86** [comment]
Let's continue in discussions.

**Comment by utack** [comment, marked as answer]
I managed to solve it
Not sure if you want the input data to reproduce it but plain and simple the original suspicion the cache was too small appears to be true


```
diff --git a/mapsforge-map-writer/src/main/java/org/mapsforge/map/writer/MapFileWriter.java b/mapsforge-map-writer/src/main/java/org/mapsforge/map/writer/MapFileWriter.java
index b381e490..36b02162 100644
--- a/mapsforge-map-writer/src/main/java/org/mapsforge/map/writer/MapFileWriter.java
+++ b/mapsforge-map-writer/src/main/java/org/mapsforge/map/writer/MapFileWriter.java
@@ -270 +270 @@ public final class MapFileWriter {
-    static final int TILE_BUFFER_SIZE = 0xA00000; // 10MB
+    static final int TILE_BUFFER_SIZE = 0xF00000; // 15MB
@@ -278 +278 @@ public final class MapFileWriter {
-    static final int WAY_DATA_BUFFER_SIZE = 0xA00000; // 10MB
+    static final int WAY_DATA_BUFFER_SIZE = 0xF00000; // 15MB
```

**Reply by devemux86** [reply]
@utack thanks for the feedback.

I don't know if changing the default buffer sizes can cause a problem in map creation at server,
as we have not seen anything yet, except https://github.com/mapsforge/mapsforge/discussions/1422 but it was reproducible with original OSM data.

We could include some buffer sizes as options in map-writer,
so users can select whatever they like for their custom maps.
